[elevation Profile] Correctly add a layer generated by a processing

ptitjano opened the pull request:
## Issue

This is somewhat similar to #59809

1. Create a new project
2. Open a new elevation profile
3. Launch any processing to create a new Layer
4. The new layer is not added to the elevation profile widget

## Solution

This is because `QgsLayerTreeRegistryBridge` listens to the `QgsProject::legendLayersAdded()` signal in order to update the elevation profile tree view. However this signal is not triggered by  the current logic in the `Postprocessing.py`.  
This PR changes the behavior of `Postprocessing.py`. by triggering the `QgsProject::legendLayersAdded()` signal and removing the duplicated node in the tree view.   

This seems like a bad solution but I'm not sure what the correct fix. Should `QgsLayerTreeRegistryBridge` listen to an other signal? Which one? Should `Postprocessing.py` be modified?   
Any thoughts?

cc  @Djedouas

nyalldawson commented:
@ptitjano 
> This seems like a bad solution but I'm not sure what the correct fix.

Indeed :laughing:  (not to mention it's only a partial fix -- if the layers are being placed in a specific group then the fix won't work either). 

It makes me nervous seeing a lot of changes to `Postprocessing.py`. This is quite a fragile file (with no unit tests at all!), and we usually get regressions whenever it's modified. (And unfortunately we can't port it to c++ for compile-time safety, because it relies on RenderingStyles and ProcessingConfig, which are currently python only). 

> Should QgsLayerTreeRegistryBridge listen to an other signal? Which one? Should Postprocessing.py be modified?

Maybe we could work around this by moving the logic postprocessing which adds the layers to the tree directly to QgsLayerTreeRegistryBridge. I.e this loop:  https://github.com/qgis/QGIS/blob/master/python/plugins/processing/gui/Postprocessing.py#L280-L296

ptitjano commented:
> Maybe we could work around this by moving the logic postprocessing which adds the layers to the tree directly to QgsLayerTreeRegistryBridge. I.e this loop: https://github.com/qgis/QGIS/blob/master/python/plugins/processing/gui/Postprocessing.py#L280-L296

Could you explain a little bit how this could be achieved please? I don't know how `QgsLayerTreeRegistryBridge` works and I can't find any example.

ptitjano committed:
postprocessing: Use QgsLayerTreeRegistryBridge to add layers

When a profile tool is already opened, the output of a procssing is
not added to its layer tree. This is because it relies on a
`QgsLayerTreeRegistryBridge`. Indeed, `QgsLayerTreeRegistryBridge`
listens to the `QgsProject::legendLayersAdded()` signal in order to
update the elevation profile tree view. However this signal is not
triggered by the current logic in `Postprocessing.py`because
`QgsProject::addMaplayer` is called with `addToLegend` set to
False. Then, the layer is added to the tree by calling
`QgsLayerTreeGroup::insertChildNode`.

This issue is fixed by creating a
`QgsLayerTreeRegistryBridge::InsertionPoint` to set the insertion
point and then calling `QgsProject::addMaplayer` with `addToLegend`
set to True.

ptitjano commented:
@lbartoletti I have completely reworked this PR. This should be ready now.

nyalldawson commented on python/plugins/processing/gui/Postprocessing.py:
```diff
@@ -289,11 +289,22 @@ def handleAlgorithmResults(
+                insertion_point = QgsLayerTreeRegistryBridge.InsertionPoint()
+
+        if insertion_point:
+            details.project.layerTreeRegistryBridge().setLayerInsertionPoint(
+                insertion_point
+            )
```
This change would need resetting afterwards, otherwise the project will behave oddly when new layers are added outside of processing.

Also, looking at how the insertion point logic works in the layer tree bridge there's a big risk of crashes! I'll try to fix those now...

ptitjano replied:
> This change would need resetting afterwards, otherwise the project will behave oddly when new layers are added outside of processing.

I have added that. However, I have realized that there is no `LayerInsertionPoint` getter. So, I have added one.

ptitjano replied:
> Also, looking at how the insertion point logic works in the layer tree bridge there's a big risk of crashes! I'll try to fix those now...

Do you mean https://github.com/qgis/QGIS/pull/60446 ?

ptitjano committed:
qgslayertreeregistrybridge: Add a layerInsertionPoint getter

nirvn commented:
@ptitjano , @nyalldawson , this has caused a regression over here whereas a model with multiple outputs will only ever insert one output into the layer tree.

nirvn commented:
(you can check the failure to add multiple output easily by running the check validity alg and witness missing output layers :))

ptitjano commented:
> @ptitjano , @nyalldawson , this has caused a regression over here whereas a model with multiple outputs will only ever insert one output into the layer tree.

Thanks for the report. I will try to take a look at it.

ptitjano commented:
> @ptitjano , @nyalldawson , this has caused a regression over here whereas a model with multiple outputs will only ever insert one output into the layer tree.

I did some tests with the 'check validity' algorithm and I'm unable to reproduce the issue. The 3 outputs are correctly displayed in the layer tree.

Maybe, this only happens on some specific layer tree configuration? Could you please share more information to reproduce the issue?

agiudiceandrea commented:
Please have a look at the regression reported at https://github.com/qgis/QGIS/issues/61146.
